### PR TITLE
PHP: Adding a new error hint, as suggested in #5078.

### PR DESCRIPTION
**Pull request by troizet**
Adding a new hint that checks if the return statement has a value if the return type is specified for the function, as suggested in #5078.

### CODE REVIEW

File: php/php.editor/src/org/netbeans/modules/php/editor/verification/ReturnStatementWithoutValueHintError.java [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+
+    @NbBundle.Messages({
+        "ReturnStatementWithoutValueHintErrorDesc=Return statement must be filled with the value"
+    })
+    private void checkReturnStatementsWithoutValue(Set<ReturnStatement> statements, String name, List<Hint> hints) {
+        if (!Type.VOID.equals(name) && !isNeverType(name)) {
+            // check empty return statement
+            statements.forEach((statement) -> {
+                if (CancelSupport.getDefault().isCancelled()) {
+                    return;
```
Comment (lbownik) [inline php/php.editor/src/org/netbeans/modules/php/editor/verification/ReturnStatementWithoutValueHintError.java:135]: this will return from lambda, not from checkReturnStatementsWithoutValue
is it intentional?

Comment (troizet) [reply php/php.editor/src/org/netbeans/modules/php/editor/verification/ReturnStatementWithoutValueHintError.java:135]: I got this code from here [php/php.editor/src/org/netbeans/modules/php/editor/verification/ReturnTypeHintError.java](https://github.com/apache/netbeans/blob/cef11a8c61656b7e15562a9d4c28e9d386a1d7f1/php/php.editor/src/org/netbeans/modules/php/editor/verification/ReturnTypeHintError.java#L135-L148).
Is it wrong?

Comment (lbownik) [reply php/php.editor/src/org/netbeans/modules/php/editor/verification/ReturnStatementWithoutValueHintError.java:135]: I don't know. I just spotted a strange costruct and brought it to your attention. Maybe it is perfectly fine.

Comment (mbien) [reply php/php.editor/src/org/netbeans/modules/php/editor/verification/ReturnStatementWithoutValueHintError.java:135]: good catch. This looks like a little cancelability bug. Both cases should be reverted to normal loops. This is probably from the time when NB had a hint which recommended to use the for each loop - and someone did this by accident and/or blindly. I turned that hint off via #3300.

If you want you can simply add another commit to this PR which fixes this and we merge this unsquashed. Might be worth checking if this happens in other hints too.

Comment (mbien) [reply php/php.editor/src/org/netbeans/modules/php/editor/verification/ReturnStatementWithoutValueHintError.java:135]: this isn't bad performance wise since it would just loop without doing anything when canceled. But this can introduce bugs later e.g if someone adds something after the `foreach` loop.

Comment (mbien) [reply php/php.editor/src/org/netbeans/modules/php/editor/verification/ReturnStatementWithoutValueHintError.java:135]: on second thought, this is so minor that you can simply add this to your first commit. feel free to force push if you find other occurrences :)

Comment (troizet) [reply php/php.editor/src/org/netbeans/modules/php/editor/verification/ReturnStatementWithoutValueHintError.java:135]: Fixed.

**Comment by junichi11** [comment]
@troizet Please implement it in `ReturnTypeHintError`.

**Comment by troizet** [comment]
That's what I did at first. Then I decided it would be easier to do it separately. OK, I'll try again. Thank you!

File: php/php.editor/src/org/netbeans/modules/php/editor/verification/ReturnTypeHintError.java [hunk 2]
```diff
@@ -116,13 +118,18 @@ private void checkReturnType(Map<ASTNode, Set<ReturnStatement>> returnStatements
...
+                checkReturnStatementsWithoutValue(statements, name, hints);
             } else if (returnType instanceof NullableType) {
                 Expression type = ((NullableType) returnType).getType();
                 if (type instanceof NamespaceName) {
                     NamespaceName namespaceName = (NamespaceName) type;
                     String name = CodeUtils.extractUnqualifiedName(namespaceName);
                     checkInvalidVoidAndNeverReturnType(type, name, hints);
-                }
+                } 
+                checkReturnStatementsWithoutValue(statements, "", hints);
```
Comment (junichi11) [inline php/php.editor/src/org/netbeans/modules/php/editor/verification/ReturnTypeHintError.java:129]: `// NOI18N`

File: php/php.editor/src/org/netbeans/modules/php/editor/verification/ReturnTypeHintError.java [hunk 3]
```diff
@@ -116,13 +118,18 @@ private void checkReturnType(Map<ASTNode, Set<ReturnStatement>> returnStatements
...
+            } else if (returnType instanceof UnionType
+                    || returnType instanceof IntersectionType) {
+                checkReturnStatementsWithoutValue(statements, "", hints);
```
Comment (junichi11) [inline php/php.editor/src/org/netbeans/modules/php/editor/verification/ReturnTypeHintError.java:132]: `// NOI18N`

File: php/php.editor/test/unit/data/testfiles/verification/ReturnTypeHintError/testReturnStatementWithoutValueHintError.php [hunk 4]
```diff
@@ -0,0 +1,106 @@
+<?php
```
Comment (junichi11) [inline php/php.editor/test/unit/data/testfiles/verification/ReturnTypeHintError/testReturnStatementWithoutValueHintError.php:1]: Please add the license header.

**Comment by junichi11** [comment]
Please fix the commit message and the title of this PR. Thanks!

File: php/php.editor/src/org/netbeans/modules/php/editor/verification/ReturnTypeHintError.java [hunk 5]
```diff
@@ -116,13 +118,18 @@ private void checkReturnType(Map<ASTNode, Set<ReturnStatement>> returnStatements
...
                 checkVoidAndNeverReturnStatements(statements, name, hints);
+                checkReturnStatementsWithoutValue(statements, name, hints);
             } else if (returnType instanceof NullableType) {
                 Expression type = ((NullableType) returnType).getType();
                 if (type instanceof NamespaceName) {
                     NamespaceName namespaceName = (NamespaceName) type;
                     String name = CodeUtils.extractUnqualifiedName(namespaceName);
                     checkInvalidVoidAndNeverReturnType(type, name, hints);
-                }
+                }
```
Comment (tmysik) [inline php/php.editor/src/org/netbeans/modules/php/editor/verification/ReturnTypeHintError.java:128]: Remove the trailing whitespace.

File: php/php.editor/src/org/netbeans/modules/php/editor/verification/ReturnTypeHintError.java [hunk 6]
```diff
@@ -133,17 +140,17 @@ private void checkReturnType(Map<ASTNode, Set<ReturnStatement>> returnStatements
         "ReturnTypeHintErrorVoidDesc=\"{0}\" cannot return anything"
     })
     private void checkVoidAndNeverReturnStatements(Set<ReturnStatement> statements, String name, List<Hint> hints) {
-        if (Type.VOID.equals(name) || isNeverType(name)) {
+        if (isVoidType(name) || isNeverType(name)) {
             // check empty return statement
-            statements.forEach((statement) -> {
+            for (ReturnStatement statement: statements) {
```
Comment (tmysik) [inline php/php.editor/src/org/netbeans/modules/php/editor/verification/ReturnTypeHintError.java:145]: Why this change?

Comment (tmysik) [reply php/php.editor/src/org/netbeans/modules/php/editor/verification/ReturnTypeHintError.java:145]: I think I see now - because of the `CancelSupport`, right?

File: php/php.editor/src/org/netbeans/modules/php/editor/verification/ReturnTypeHintError.java [hunk 7]
```diff
@@ -152,17 +159,40 @@ private void checkVoidAndNeverReturnStatements(Set<ReturnStatement> statements,
...
+    }
+
     private boolean isNeverType(String name) {
         return getPhpVersion(fileObject).hasNeverType()
                 && Type.NEVER.equals(name);
     }
 
+   private boolean isVoidType(String name) {
+        return getPhpVersion(fileObject).hasVoidReturnType()
+                && Type.VOID.equals(name);
```
Comment (tmysik) [inline php/php.editor/src/org/netbeans/modules/php/editor/verification/ReturnTypeHintError.java:193]: I am not sure, but I would expect that if we swap the conditions, the performance should be better (comparing strings should be faster than working with file objects).

Comment (tmysik) [reply php/php.editor/src/org/netbeans/modules/php/editor/verification/ReturnTypeHintError.java:193]: (It applies also for `isNeverType` method.)

**Review by tmysik — APPROVED** [review]
Thanks for your work!

File: php/php.editor/src/org/netbeans/modules/php/editor/verification/ReturnTypeHintError.java (hunk 2 above)
Comment (troizet) [reply php/php.editor/src/org/netbeans/modules/php/editor/verification/ReturnTypeHintError.java]: Done.

File: php/php.editor/src/org/netbeans/modules/php/editor/verification/ReturnTypeHintError.java (hunk 3 above)
Comment (troizet) [reply php/php.editor/src/org/netbeans/modules/php/editor/verification/ReturnTypeHintError.java]: Done.

File: php/php.editor/test/unit/data/testfiles/verification/ReturnTypeHintError/testReturnStatementWithoutValueHintError.php (hunk 4 above)
Comment (troizet) [reply php/php.editor/test/unit/data/testfiles/verification/ReturnTypeHintError/testReturnStatementWithoutValueHintError.php:1]: Done.

File: php/php.editor/src/org/netbeans/modules/php/editor/verification/ReturnTypeHintError.java (hunk 5 above)
Comment (troizet) [reply php/php.editor/src/org/netbeans/modules/php/editor/verification/ReturnTypeHintError.java]: Done.

File: php/php.editor/src/org/netbeans/modules/php/editor/verification/ReturnTypeHintError.java (hunk 7 above)
Comment (troizet) [reply php/php.editor/src/org/netbeans/modules/php/editor/verification/ReturnTypeHintError.java]: Done.

File: php/php.editor/src/org/netbeans/modules/php/editor/verification/ReturnTypeHintError.java (hunk 6 above)
Comment (troizet) [reply php/php.editor/src/org/netbeans/modules/php/editor/verification/ReturnTypeHintError.java:145]: Yes, you are right.

**Review by junichi11 — APPROVED** [review]
Thanks for your work.

**Comment by tmysik** [comment]
@junichi11 Not sure whether we need all the approvals; if not, feel free to merge anytime. Thank you.

**Comment by troizet** [comment]
Thanks to all!
